File: swarm/app/llm/groq_client.py
"""Groq LLM client — multi-key pool, 40 req/min sliding-window rate limiter.

Mirrors the architecture of nim_client.py so llm_dispatcher.py can swap
providers transparently.

Key management:
  - Reads GROQ_API_KEY, GROQ_API_KEY_2 … GROQ_API_KEY_20 from env
  - Dead-key kill list (401/403 — skip for process lifetime)
  - Least-used-first key selection via sliding 60s window counter
  - On 429: marks key as rate-limited for remainder of window, rotates instantly
  - AllGroqKeysExhaustedError raised when all keys are at limit/dead → caller falls back to NIM

Model routing (via config.py):
  - conversational: llama-3.3-70b-versatile   (Scholar chat)
  - routing:        llama-3.1-8b-instant       (intent classification, summarizer)
  - structured:     NOT used — NIM Mixtral-8x22B handles that tier
"""

from __future__ import annotations

import asyncio
import json as _json
import logging
import os
import re
import time
from collections import deque
from collections.abc import AsyncIterator
from dataclasses import dataclass, field

import httpx

from app.config import GenerationProfile, get_settings

logger = logging.getLogger("aroha.groq")

# ---------------------------------------------------------------------------
# Process-lifetime key state
# ---------------------------------------------------------------------------

_dead_keys: dict[str, str] = {}          # key → reason
_rate_windows: dict[str, deque] = {}     # key → deque of UNIX timestamps (last 60s)
_MAX_NUMBERED_SLOT = 20
_WINDOW_SECONDS = 60


def _sanitize_key(key: str) -> str:
    return key.replace("\ufeff", "").strip()


def get_available_keys() -> list[str]:
    """Collect all configured Groq keys, excluding dead ones."""
    collected: list[str] = []

    def push_if_set(raw: str | None) -> None:
        if not raw:
            return
        clean = _sanitize_key(raw)
        if clean:
            collected.append(clean)

    push_if_set(os.environ.get("GROQ_API_KEY"))
    for i in range(2, _MAX_NUMBERED_SLOT + 1):
        push_if_set(os.environ.get(f"GROQ_API_KEY_{i}"))

    all_keys = list(dict.fromkeys(collected))  # deduplicate, preserve order
    live = [k for k in all_keys if k not in _dead_keys]

    if not live and all_keys:
        logger.warning("All Groq keys were marked dead — resetting kill list")
        _dead_keys.clear()
        return all_keys
    return live


def _requests_this_window(key: str) -> int:
    """Count requests made on this key in the last 60 seconds."""
    now = time.monotonic()
    if key not in _rate_windows:
        _rate_windows[key] = deque()
    dq = _rate_windows[key]
    # Evict expired timestamps
    while dq and now - dq[0] > _WINDOW_SECONDS:
        dq.popleft()
    return len(dq)


def _record_request(key: str) -> None:
    """Record a new request timestamp for this key."""
    if key not in _rate_windows:
        _rate_windows[key] = deque()
    _rate_windows[key].append(time.monotonic())


def _is_rate_limited(key: str) -> bool:
    """Return True if this key has hit the configured RPM limit."""
    rpm_limit = get_settings().groq_rpm_limit
    return _requests_this_window(key) >= rpm_limit


def _pick_key(keys: list[str]) -> str | None:
    """Pick the least-used live key that hasn't hit its RPM limit."""
    candidates = [
        (k, _requests_this_window(k))
        for k in keys
        if k not in _dead_keys and not _is_rate_limited(k)
    ]
    if not candidates:
        return None
    # Least-busy first
    candidates.sort(key=lambda x: x[1])
    return candidates[0][0]


# ---------------------------------------------------------------------------
# Error types
# ---------------------------------------------------------------------------

@dataclass
class GroqError(Exception):
    message: str
    status: int | None = None

    def __str__(self) -> str:
        return self.message


class AllGroqKeysExhaustedError(GroqError):
    """Raised when every configured key is dead or rate-limited this window."""
    pass


def _is_dead_key_error(status: int | None, msg: str) -> bool:
    if status in (401, 403):
        return True
    if re.search(r"invalid\s*api\s*key", msg, re.I):
        return True
    if re.search(r"api\s*key.*(expired|revoked|disabled)", msg, re.I):
        return True
    return False


# ---------------------------------------------------------------------------
# Model resolution
# ---------------------------------------------------------------------------

def _resolve_model(profile: GenerationProfile) -> str:
    s = get_settings()
    if profile.model_tier == "conversational":
        return s.groq_model_conversational
    # routing tier and anything else → fast 8B
    return s.groq_model_routing


# ---------------------------------------------------------------------------
# Core single-call helpers
# ---------------------------------------------------------------------------

async def _call_groq_once(
    client: httpx.AsyncClient,
    api_key: str,
    model: str,
    messages: list[dict],
    profile: GenerationProfile,
) -> dict:
    settings = get_settings()
    body: dict = {
        "model": model,
        "messages": messages,
        "max_tokens": profile.max_tokens,
        "temperature": profile.temperature,
        "stream": False,
    }

    resp = await client.post(
        f"{settings.groq_base_url}/chat/completions",
        json=body,
        headers={"Authorization": f"Bearer {api_key}"},
        timeout=60.0,
    )

    if resp.status_code != 200:
        err_text = resp.text
        raise GroqError(
            message=f"Groq error {resp.status_code}: {err_text[:300]}",
            status=resp.status_code,
        )

    return resp.json()


async def _stream_groq(
    client: httpx.AsyncClient,
    api_key: str,
    model: str,
    messages: list[dict],
    profile: GenerationProfile,
) -> AsyncIterator[str]:
    settings = get_settings()
    body: dict = {
        "model": model,
        "messages": messages,
        "max_tokens": profile.max_tokens,
        "temperature": profile.temperature,
        "stream": True,
    }

    async with client.stream(
        "POST",
        f"{settings.groq_base_url}/chat/completions",
        json=body,
        headers={"Authorization": f"Bearer {api_key}"},
        timeout=60.0,
    ) as resp:
        if resp.status_code != 200:
            err_text = await resp.aread()
            raise GroqError(
                message=f"Groq error {resp.status_code}: {err_text.decode()[:300]}",
                status=resp.status_code,
            )
        async for line in resp.aiter_lines():
            if not line.startswith("data: "):
                continue
            payload = line[6:].strip()
            if payload == "[DONE]":
                return
            try:
                chunk = _json.loads(payload)
            except _json.JSONDecodeError:
                continue
            delta = chunk.get("choices", [{}])[0].get("delta", {})
            content = delta.get("content")
            if content:
                yield content


# ---------------------------------------------------------------------------
# Key-rotating wrapper
# ---------------------------------------------------------------------------

async def _with_key_rotation(
    keys: list[str],
    model: str,
    messages: list[dict],
    profile: GenerationProfile,
    *,
    streaming: bool = False,
):
    """Try each available key in least-busy order. Raise AllGroqKeysExhaustedError if all fail."""
    if not keys:
        raise AllGroqKeysExhaustedError("No Groq API keys configured")

    async with httpx.AsyncClient() as client:
        # Try up to len(keys) times, rotating on rate-limit or transient errors
        attempts = 0
        max_attempts = len(keys) * 2  # allow retrying after brief wait

        while attempts < max_attempts:
            key = _pick_key(keys)
            if key is None:
                # All keys rate-limited — wait until the oldest window entry expires
                logger.warning(
                    "All %d Groq keys at RPM limit — waiting for window to clear", len(keys)
                )
                raise AllGroqKeysExhaustedError(
                    f"All {len(keys)} Groq keys have hit the {get_settings().groq_rpm_limit} req/min limit"
                )

            _record_request(key)
            attempts += 1

            try:
                if streaming:
                    return _stream_groq(client, key, model, messages, profile)
                else:
                    return await _call_groq_once(client, key, model, messages, profile)

            except GroqError as err:
                if _is_dead_key_error(err.status, err.message):
                    _dead_keys[key] = err.message[:100]
                    logger.warning(
                        "Groq key ...%s marked DEAD: %s", key[-6:], err.message[:100]
                    )
                    continue

                if err.status == 429:
                    # Force this key to look fully used for this window
                    logger.warning(
                        "Groq key ...%s returned 429 — rotating to next key", key[-6:]
                    )
                    rpm_limit = get_settings().groq_rpm_limit
                    now = time.monotonic()
                    dq = _rate_windows.setdefault(key, deque())
                    while len(dq) < rpm_limit:
                        dq.append(now)
                    continue

                logger.warning(
                    "Groq key ...%s error (status=%s): %s — rotating",
                    key[-6:], err.status, err.message[:80],
                )
                continue

            except Exception as err:
                logger.warning("Groq key ...%s exception: %s — rotating", key[-6:], err)
                continue

    raise AllGroqKeysExhaustedError(
        f"All Groq keys failed after {attempts} attempts"
    )


# ---------------------------------------------------------------------------
# Public API (same signatures as nim_client for drop-in use via dispatcher)
# ---------------------------------------------------------------------------

async def generate(
    messages: list[dict],
    profile: GenerationProfile,
    *,
    system: str | None = None,
) -> str:
    """Buffered generation. Returns the full response text."""
    keys = get_available_keys()
    model = _resolve_model(profile)

    if system:
        messages = [{"role": "system", "content": system}, *messages]

    data = await _with_key_rotation(keys, model, messages, profile, streaming=False)
    return (
        data.get("choices", [{}])[0]
        .get("message", {})
        .get("content", "")
    )


async def stream(
    messages: list[dict],
    profile: GenerationProfile,
    *,
    system: str | None = None,
) -> AsyncIterator[str]:
    """Streaming generation. Yields token deltas."""
    keys = get_available_keys()
    model = _resolve_model(profile)

    if system:
        messages = [{"role": "system", "content": system}, *messages]

    result = await _with_key_rotation(keys, model, messages, profile, streaming=True)
    async for token in result:
        yield token

File: src/lib/llm/groq-client.ts
// =============================================================================
// Groq LLM Client
// Multi-key pool with 40 req/min sliding-window rate limiter per key,
// dead-key cooldown, least-busy-first selection, and exponential backoff.
// Mirrors the nim-client architecture.
// =============================================================================

import { env } from '../../config/env.js';
import { type LLMRequestOptions } from '../../config/llm.js';
import { logger } from '../logger.js';

// =============================================================================
// Error Classes
// =============================================================================

export class GroqError extends Error {
  constructor(
    message: string,
    public readonly statusCode?: number,
    public readonly body?: string,
  ) {
    super(message);
    this.name = 'GroqError';
  }
}

export class AllGroqKeysExhaustedError extends GroqError {
  constructor(message = 'All Groq API keys are exhausted, rate-limited, or dead') {
    super(message);
    this.name = 'AllGroqKeysExhaustedError';
  }
}

// =============================================================================
// Key Pool & Rate Limiting
// =============================================================================

interface GroqKeyState {
  key: string;
  /** Epoch ms until which this key is considered dead; 0 = alive. */
  deadUntil: number;
  /** Timestamps (epoch ms) of requests made in the last 60 seconds. */
  requestWindow: number[];
}

const DEAD_KEY_COOLDOWN_MS = 5 * 60_000;
const WINDOW_MS = 60_000;

function loadGroqKeys(): GroqKeyState[] {
  const keys: GroqKeyState[] = [];

  if (env.GROQ_API_KEY) {
    keys.push({ key: env.GROQ_API_KEY, deadUntil: 0, requestWindow: [] });
  }

  for (let i = 2; i <= 20; i++) {
    const envKey = `GROQ_API_KEY_${i}` as keyof typeof env;
    const value = env[envKey] as string | undefined;
    if (value) {
      keys.push({ key: value, deadUntil: 0, requestWindow: [] });
    }
  }

  return keys;
}

const keyPool: GroqKeyState[] = loadGroqKeys();

function markDead(keyState: GroqKeyState): void {
  keyState.deadUntil = Date.now() + DEAD_KEY_COOLDOWN_MS;
}

function cleanWindow(keyState: GroqKeyState, now: number): void {
  const cutoff = now - WINDOW_MS;
  keyState.requestWindow = keyState.requestWindow.filter((t) => t > cutoff);
}

function isRateLimited(keyState: GroqKeyState, now: number): boolean {
  cleanWindow(keyState, now);
  return keyState.requestWindow.length >= env.GROQ_RPM_LIMIT;
}

function recordRequest(keyState: GroqKeyState, now: number): void {
  keyState.requestWindow.push(now);
}

// =============================================================================
// Dead-Key Detection
// =============================================================================

function isDeadKeyError(status: number, msg: string): boolean {
  if (status === 401 || status === 403) return true;
  if (/invalid\s*api\s*key/i.test(msg)) return true;
  if (/api\s*key.*(expired|revoked|disabled)/i.test(msg)) return true;
  return false;
}

// =============================================================================
// Key Selection
// =============================================================================

function selectKey(): GroqKeyState | null {
  const now = Date.now();
  const candidates = keyPool.filter((k) => k.deadUntil <= now && !isRateLimited(k, now));
  if (candidates.length === 0) return null;
  // Least-busy first
  candidates.sort((a, b) => a.requestWindow.length - b.requestWindow.length);
  return candidates[0] ?? null;
}

// =============================================================================
// Helpers
// =============================================================================

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/** Backoff when ALL keys are simultaneously rate-limited (no key to switch to). */
function rateLimitBackoff(rateLimitWaits: number): number {
  return Math.min(1000 * Math.pow(2, rateLimitWaits), 10_000);
}

/**
 * Cap on how long we'll ever actually sleep for a single 429, REGARDLESS of
 * what Groq's Retry-After header says. Groq has returned Retry-After values
 * of 20-30+ minutes (real quota exhaustion, not a short RPM window) — honoring
 * that verbatim leaves the SSE stream silent long enough for nginx's
 * proxy_read_timeout to kill the connection outright. Capping means we fail
 * fast (and cleanly, via the normal error path) instead of hanging.
 */
const MAX_RATE_LIMIT_WAIT_MS = 10_000;

const GENERATE_TIMEOUT_MS = 60_000;
const STREAM_TIMEOUT_MS = 120_000;
/**
 * 429s don't consume the real (network-error/dead-key) attempt budget — they
 * back off and retry instead, up to this many rounds, since with only a
 * handful of keys a burst of concurrent requests (chat + horoscope + chat
 * summary + purchase-plan all share this pool) can 429 every key at once
 * without any of them being individually "dead". Bounded well under nginx's
 * proxy timeout so a real failure still surfaces instead of hanging forever.
 */
const MAX_RATE_LIMIT_RETRIES = 12;

function makeAbort(external: AbortSignal | undefined, ms: number) {
  const ac = new AbortController();
  const onExternal = () => ac.abort();
  if (external) {
    if (external.aborted) ac.abort();
    else external.addEventListener('abort', onExternal, { once: true });
  }
  const timer = setTimeout(() => ac.abort(), ms);
  return {
    signal: ac.signal,
    clear: () => {
      clearTimeout(timer);
      external?.removeEventListener('abort', onExternal);
    },
  };
}

function resolveModel(opts: LLMRequestOptions): string {
  if (opts.model) return opts.model;
  if (opts.profile.modelTier === 'structured') return env.GROQ_MODEL_STRUCTURED;
  if (opts.profile.modelTier === 'conversational') return env.GROQ_MODEL_CONVERSATIONAL;
  return env.GROQ_MODEL_ROUTING;
}

// =============================================================================
// Core Request
// =============================================================================

async function doRequest(
  keyState: GroqKeyState,
  opts: LLMRequestOptions,
  signal: AbortSignal,
): Promise<Response> {
  const model = resolveModel(opts);
  const baseUrl = env.GROQ_BASE_URL;

  const body: Record<string, unknown> = {
    model,
    messages: opts.messages,
    temperature: opts.profile.temperature,
    max_tokens: opts.profile.maxTokens,
    stream: opts.profile.stream,
  };

  if (opts.profile.jsonMode) {
    body.response_format = { type: 'json_object' };
  }

  return fetch(`${baseUrl}/chat/completions`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${keyState.key}`,
    },
    body: JSON.stringify(body),
    signal,
  });
}

// =============================================================================
// Buffered Generate (non-streaming)
// =============================================================================

export async function generate(opts: LLMRequestOptions): Promise<string> {
  if (keyPool.length === 0) {
    throw new AllGroqKeysExhaustedError('No Groq API keys configured');
  }

  const MAX_ATTEMPTS = keyPool.length * 2;
  let attempts = 0;
  let rateLimitWaits = 0;

  while (attempts < MAX_ATTEMPTS) {
    const keyState = selectKey();
    if (!keyState) {
      if (rateLimitWaits >= MAX_RATE_LIMIT_RETRIES) {
        throw new AllGroqKeysExhaustedError(
          `All Groq keys rate-limited after ${MAX_RATE_LIMIT_RETRIES} backoff rounds`,
        );
      }
      const waitMs = rateLimitBackoff(rateLimitWaits);
      rateLimitWaits++;
      logger.warn({ waitMs, rateLimitWaits }, 'All Groq keys rate-limited, backing off');
      await sleep(waitMs);
      continue;
    }

    const now = Date.now();
    recordRequest(keyState, now);
    attempts++;

    const abort = makeAbort(opts.signal, opts.timeoutMs ?? GENERATE_TIMEOUT_MS);
    let response: Response;
    let bodyText: string;

    try {
      const nonStreamProfile = { ...opts.profile, stream: false };
      response = await doRequest(keyState, { ...opts, profile: nonStreamProfile }, abort.signal);
      bodyText = await response.text();
    } catch (err) {
      logger.warn({ err, attempt: attempts }, 'Groq request network error/timeout');
      if (attempts < MAX_ATTEMPTS) {
        continue;
      }
      throw new GroqError(`Network error after ${MAX_ATTEMPTS} attempts: ${String(err)}`);
    } finally {
      abort.clear();
    }

    if (isDeadKeyError(response.status, bodyText)) {
      logger.warn({ status: response.status }, 'Groq key marked dead (cooldown)');
      markDead(keyState);
      continue;
    }

    if (response.status === 429) {
      // Max out the window so this key won't be selected again for a minute.
      while (keyState.requestWindow.length < env.GROQ_RPM_LIMIT) {
        keyState.requestWindow.push(Date.now());
      }
      if (rateLimitWaits >= MAX_RATE_LIMIT_RETRIES) {
        throw new AllGroqKeysExhaustedError(
          `Rate limited after ${MAX_RATE_LIMIT_RETRIES} backoff rounds`,
        );
      }
      const retryAfterHeader = response.headers.get('Retry-After');
      const retryAfterSec = retryAfterHeader ? parseInt(retryAfterHeader, 10) : NaN;
      const waitMs = Number.isNaN(retryAfterSec)
        ? rateLimitBackoff(rateLimitWaits)
        : Math.min(retryAfterSec * 1000, MAX_RATE_LIMIT_WAIT_MS);
      rateLimitWaits++;
      logger.warn(
        { status: response.status, waitMs, rateLimitWaits },
        'Groq 429 rate limited, backing off',
      );
      await sleep(waitMs);
      // Doesn't consume the real attempt budget — only genuine network/dead-key
      // failures should burn through MAX_ATTEMPTS.
      attempts--;
      continue;
    }

    if (!response.ok) {
      logger.warn({ status: response.status, body: bodyText.slice(0, 500) }, 'Groq API error');
      if (attempts < MAX_ATTEMPTS) {
        continue;
      }
      throw new GroqError(
        `Groq API error ${response.status}: ${bodyText.slice(0, 500)}`,
        response.status,
        bodyText,
      );
    }

    let data: any;
    try {
      data = JSON.parse(bodyText);
    } catch (parseErr) {
      logger.warn({ sample: bodyText.slice(0, 500) }, 'Groq success body not valid JSON');
      throw new GroqError(
        `Groq returned non-JSON success body: ${String(parseErr)}`,
        response.status,
      );
    }

    const content = data.choices?.[0]?.message?.content ?? '';
    return content;
  }

  throw new AllGroqKeysExhaustedError(`Exhausted all ${MAX_ATTEMPTS} attempts on Groq`);
}

// =============================================================================
// Streaming Generate (SSE async generator)
// =============================================================================

export async function* stream(opts: LLMRequestOptions): AsyncGenerator<string, void, unknown> {
  if (keyPool.length === 0) {
    throw new AllGroqKeysExhaustedError('No Groq API keys configured');
  }

  const MAX_ATTEMPTS = keyPool.length * 2;
  let attempts = 0;
  let rateLimitWaits = 0;
  let yieldedAny = false;

  while (attempts < MAX_ATTEMPTS) {
    const keyState = selectKey();
    if (!keyState) {
      if (rateLimitWaits >= MAX_RATE_LIMIT_RETRIES) {
        throw new AllGroqKeysExhaustedError(
          `All Groq keys rate-limited after ${MAX_RATE_LIMIT_RETRIES} backoff rounds`,
        );
      }
      const waitMs = rateLimitBackoff(rateLimitWaits);
      rateLimitWaits++;
      logger.warn({ waitMs, rateLimitWaits }, 'All Groq keys rate-limited (stream), backing off');
      await sleep(waitMs);
      continue;
    }

    const now = Date.now();
    recordRequest(keyState, now);
    attempts++;

    const abort = makeAbort(opts.signal, STREAM_TIMEOUT_MS);
    try {
      const streamProfile = { ...opts.profile, stream: true };
      let response: Response;
      try {
        response = await doRequest(keyState, { ...opts, profile: streamProfile }, abort.signal);
      } catch (err) {
        logger.warn({ err, attempt: attempts }, 'Groq stream request network error/timeout');
        if (yieldedAny) throw err;
        continue;
      }

      if (!response.ok) {
        const bodyText = await response.text();
        if (isDeadKeyError(response.status, bodyText)) {
          logger.warn({ status: response.status }, 'Groq key marked dead (cooldown)');
          markDead(keyState);
          if (yieldedAny)
            throw new GroqError('Groq key died mid-stream', response.status, bodyText);
          continue;
        }

        if (response.status === 429) {
          while (keyState.requestWindow.length < env.GROQ_RPM_LIMIT) {
            keyState.requestWindow.push(Date.now());
          }
          if (yieldedAny) throw new GroqError('Groq 429 mid-stream', 429, bodyText);
          if (rateLimitWaits >= MAX_RATE_LIMIT_RETRIES) {
            throw new AllGroqKeysExhaustedError(
              `Rate limited after ${MAX_RATE_LIMIT_RETRIES} backoff rounds`,
            );
          }
          const retryAfterHeader = response.headers.get('Retry-After');
          const retryAfterSec = retryAfterHeader ? parseInt(retryAfterHeader, 10) : NaN;
          const waitMs = Number.isNaN(retryAfterSec)
            ? rateLimitBackoff(rateLimitWaits)
            : Math.min(retryAfterSec * 1000, MAX_RATE_LIMIT_WAIT_MS);
          rateLimitWaits++;
          logger.warn(
            { status: response.status, waitMs, rateLimitWaits },
            'Groq 429 stream rate limited, backing off',
          );
          await sleep(waitMs);
          attempts--;
          continue;
        }

        throw new GroqError(
          `Groq API error ${response.status}: ${bodyText.slice(0, 500)}`,
          response.status,
          bodyText,
        );
      }

      const reader = response.body?.getReader();
      if (!reader) throw new GroqError('No response body from Groq');

      const decoder = new TextDecoder();
      let buffer = '';

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split('\n');
        buffer = lines.pop() ?? '';

        for (const line of lines) {
          if (!line.startsWith('data: ')) continue;
          const data = line.slice(6).trim();
          if (data === '[DONE]') return;

          try {
            const parsed = JSON.parse(data);
            const content = parsed.choices?.[0]?.delta?.content;
            if (content) {
              yieldedAny = true;
              yield content;
            }
          } catch (e) {
            // ignore malformed chunks
          }
        }
      }
      return;
    } finally {
      abort.clear();
    }
  }

  throw new AllGroqKeysExhaustedError(`Exhausted all ${MAX_ATTEMPTS} stream attempts on Groq`);
}
